File: VocaDbModel/Helpers/SongHelper.cs
using System.Linq;
using VocaDb.Model.Domain;
using VocaDb.Model.Domain.Songs;
using VocaDb.Model.Domain.Tags;

namespace VocaDb.Model.Helpers
{
	public static class SongHelper
	{
		public static Tag[] GetGenreTags(SongInAlbum songInAlbum)
		{
			Tag[] genres;

			if (songInAlbum.Song != null)
			{
				genres = songInAlbum.Song.Tags.TagsByVotes.Where(t => t.CategoryName == TagCommonCategoryNames.Genres).ToArray();

				if (genres.Any())
					return genres.ToArray();
			}

			genres = songInAlbum.Album.Tags.TagsByVotes.Where(t => t.CategoryName == TagCommonCategoryNames.Genres).ToArray();
			return genres.ToArray();
		}

		public static ContentFocus GetContentFocus(SongType songType) => songType switch
		{
			SongType.DramaPV or SongType.MusicPV => ContentFocus.Video,
			SongType.Illustration => ContentFocus.Illustration,
			_ => ContentFocus.Music,
		};
	}
}

File: VocaDbWeb/Scripts/Helpers/SongHelper.ts
import ContentFocus from '@Models/ContentFocus';
import SongType from '@Models/Songs/SongType';
import _ from 'lodash';

export default class SongHelper {
	public static originalVersionTypes = [
		SongType.Unspecified,
		SongType.Original,
		SongType.Remaster,
		SongType.Remix,
		SongType.Cover,
		SongType.Mashup,
		SongType.DramaPV,
		SongType.Other,
	];

	public static originalVersionTypesString = (): string =>
		_.map(SongHelper.originalVersionTypes, (s) => SongType[s]).join(',');

	// Checks whether a song type is to be considered animation where animators are considered as the main role
	public static getContentFocus(songType: SongType): ContentFocus {
		switch (songType) {
			case SongType.DramaPV:
			case SongType.MusicPV:
				return ContentFocus.Video;

			case SongType.Illustration:
				return ContentFocus.Illustration;

			default:
				return ContentFocus.Music;
		}
	}

	// Checks whether a song type is to be considered instrumental where the song is allowed to have no vocalists
	public static isInstrumental(songType: SongType): boolean {
		return songType === SongType.Instrumental || songType === SongType.DramaPV;
	}
}
